File: java/src/main/java/com/newnancity/newnanmanager/services/MonitorService.java
package com.newnancity.newnanmanager.services;

import com.newnancity.newnanmanager.SessionContext;
import com.newnancity.newnanmanager.exceptions.NewNanManagerException;
import com.newnancity.newnanmanager.generated.api.MonitorServiceApi;
import com.newnancity.newnanmanager.generated.model.GetMonitorStatsResponse;
import com.newnancity.newnanmanager.generated.model.HeartbeatRequest;
import com.newnancity.newnanmanager.generated.model.HeartbeatResponse;
import com.newnancity.newnanmanager.generated.model.ServerSessionResponse;

/** Server session, heartbeat and monitor statistics service. */
public final class MonitorService extends AbstractService {
    private final MonitorServiceApi api;

    public MonitorService(MonitorServiceApi api) {
        this.api = api;
    }

    public ServerSessionResponse createServerSession(int serverId, Object body) throws NewNanManagerException {
        return execute(() -> api.createServerSession(serverId, body).execute());
    }

    public SessionContext createServerSession(int serverId) throws NewNanManagerException {
        ServerSessionResponse response = createServerSession(serverId, new Object());
        return new SessionContext(response.getSessionId(), response.getSessionEpoch());
    }

    public HeartbeatResponse heartbeat(
            int serverId,
            String sessionId,
            long sessionEpoch,
            HeartbeatRequest request) throws NewNanManagerException {
        SessionContext session = new SessionContext(sessionId, sessionEpoch);
        return execute(() -> api.heartbeat(serverId, session.getId(), session.getEpoch(), request).execute());
    }

    public HeartbeatResponse heartbeat(
            int serverId,
            HeartbeatRequest request,
            SessionContext session) throws NewNanManagerException {
        return heartbeat(serverId, session.getId(), session.getEpoch(), request);
    }

    public GetMonitorStatsResponse getMonitorStats(int serverId) throws NewNanManagerException {
        return getMonitorStats(serverId, null, null, null, null);
    }

    public GetMonitorStatsResponse getMonitorStats(
            int serverId,
            Long since,
            Long duration,
            Integer limit,
            String cursor) throws NewNanManagerException {
        return execute(() -> api.getMonitorStats(serverId)
                .since(since)
                .duration(duration)
                .limit(limit)
                .cursor(cursor)
                .execute());
    }

    public MonitorServiceApi raw() {
        return api;
    }
}

File: kotlin/src/main/kotlin/com/nanmanager/bukkit/services/MonitorService.kt
package com.nanmanager.bukkit.services

import com.nanmanager.bukkit.http.HttpClient
import com.nanmanager.bukkit.SessionContext
import com.nanmanager.bukkit.models.*

/** 监控服务；生成客户端负责请求构造，facade 保留原有模型和错误边界。 */
class MonitorService(httpClient: HttpClient) : BaseService(httpClient) {
    private val api get() = httpClient.generatedApis.monitor

    fun createServerSession(serverId: Int): SessionContext {
        return generated<SessionContext, com.newnanmanager.generated.models.ServerSessionResponse> {
            api.createServerSessionWithHttpInfo(serverId, emptyMap<String, Any>())
        }
    }

    fun heartbeat(serverId: Int, request: HeartbeatRequest): HeartbeatResponse {
        return post("/api/v1/monitor/$serverId/heartbeat", request)
    }

    fun heartbeat(serverId: Int, request: HeartbeatRequest, session: SessionContext): HeartbeatResponse {
        val body = toGenerated<com.newnanmanager.generated.models.HeartbeatRequest>(request)
        return generated<HeartbeatResponse, com.newnanmanager.generated.models.HeartbeatResponse> {
            api.heartbeatWithHttpInfo(serverId, session.id, session.epoch, body)
        }
    }

    fun getMonitorStats(request: GetMonitorStatsRequest): GetMonitorStatsResponse {
        return generated<GetMonitorStatsResponse, com.newnanmanager.generated.models.GetMonitorStatsResponse> {
            api.getMonitorStatsWithHttpInfo(
                serverId = request.serverId,
                since = request.since,
                duration = request.duration,
                limit = request.limit,
                cursor = request.cursor
            )
        }
    }
}
